os/board/rtl8721csm/src/component/os/tizenrt/tizenrt_service.c

```diff
@@ -913,36 +997,38 @@ u32 _tizenrt_timerStop(_timerHandle xTimer, osdepTickType xBlockTime)
 	if (ret != -2) {
 		kmm_free(timer->work_hdl);
 		kmm_free(timer);
-		DBG_ERR("_tizenrt_stop_timer failed! ret = %d", ret);
+		DBG_ERR(" failed! ret = %d", ret);
 		return _FAIL;
 	}
 	timer->timevalue = 0;
 	timer->live = 0;
-	DBG_ERR("_tizenrt_stop_timer is Done! timer->work_hdl = %x", timer->work_hdl);
+	DBG_ERR(" is Done! ret: %d, timer->work_hdl = %d\n", ret, timer->work_hdl);
 	return _SUCCESS;
 }
 
 u32 _tizenrt_timerChangePeriod(_timerHandle xTimer, osdepTickType xNewPeriod, osdepTickType xBlockTime)
 {
 	int ret;
 	struct timer_list_priv *timer = (struct timer_list_priv *)xTimer;
-	ret = work_queue(LPWORK, timer->work_hdl, timer->function, (void *)(timer), xNewPeriod);
+	ret = work_queue(LPWORK, timer->work_hdl, _tizenrt_timer_wrapper, (void *)(timer), xNewPeriod);
 	if (ret == -EALREADY) {
 		if (work_cancel(LPWORK, timer->work_hdl) != OK) {
 			goto cleanup;
```

Comment:
If work_cancel failed then is it ok to free timer->work_hdl which might be still in lpwork?

Reply:
_tizenrt_timerChangePeriod() will try to change or start the timer, if work_queue() is fail mean timer is already running.
So work_cancel() failed should not happen.

Reply:
Thanks for explanation.

Please share your opinion of following scenario
```
taskA executed work_queue() at line 985.
work_queue returned -EALREADY
context switching
lpwork executed timer
context switching
taskA executed work_cancel() and it returned ENOENT because it already had been run
taskA cleanup timer->workhdl => access to invalid memory
```
If  above scenario is wrong, please let me know.

Reply:
Yes, situation you mention it can happen, the free of timer handler will be remove, it will direct return Fail.